docs/source/extending/overloading-guide.rst

```diff
@@ -0,0 +1,278 @@
+  NumPy arrays such as scalar, list, tuple, set, iterator, generator etc.. So
+  you will need to account for that during type inference and subsequently as
+  part of the tests.
+
+* A NumPy function may return a scalar, array or a data structure
+  which matches one of its inputs - so you need to watch out for type
+  unification problems and dispatch to appropriate implementations. For
+  example, ``np.corrcoef`` may return an array or a scalar depending on it's
+  inputs.
+
+* If you are implementing a new function, you should always update the
+  `documentation
+  <http://numba.pydata.org/numba-doc/latest/reference/numpysupported.html>`_.
+  The sources can be found in `docs/source/reference/numpysupported.rst``. Be
+  sure to mention any limitations that your implementation has, e.g. no support
+  for the ``axis`` keyword.
+
+* When writing tests for the functionality itself, it's useful to include
+  handling of non-finite values; arrays with different shapes and layouts;
+  complex inputs; scalar inputs; inputs with types for which support is not
+  documented (e.g. a function which the NumPy docs say requires a float or int
+  input might also 'work' if given a bool or complex input).
+
+* When writing tests for exceptions, for example, when adding tests to
+  ``numba/tests/test_np_functions.py`` you may encounter the following error
+  message:
+
+  .. code:: pycon
+
+        ======================================================================
+        FAIL: test_foo (numba.tests.test_np_functions.TestNPFunctions)
+        ----------------------------------------------------------------------
+        Traceback (most recent call last):
+        File "/Users/vhaenel/git/numba/numba/tests/support.py", line 645, in tearDown
+            self.memory_leak_teardown()
+        File "/Users/vhaenel/git/numba/numba/tests/support.py", line 619, in memory_leak_teardown
+            self.assert_no_memory_leak()
+        File "/Users/vhaenel/git/numba/numba/tests/support.py", line 628, in assert_no_memory_leak
+            self.assertEqual(total_alloc, total_free)
+        AssertionError: 36 != 35
+
+  This is caused because some exceptions leak references. Ideally, you will
+  place all exception testing in a separate test method and then add a call to
+  ``self.disable_leak_check()`` to disable the leak-check. (Remember to inherit
+  from ``numba.tests.support.TestCase`` to make that available).
+
+* For many of the functions that are available in NumPy, there are
+  corresponding methods defined on the NumPy array type. For example, the
+  function ``repeat`` is available in two flavours.
+
+  .. code:: python
+
+        import numpy as np
+        a = np.arange(10)
+        # function
+        np.repeat(a, 10)
+        # method
+        a.repeat(10)
+
+  Once you have written the function implementation, you can easily use
+  ``overload_method`` and reuse it, for example for the ``repeat`` function/method.
```

Comment:
End the sentence on the comma and move the `For example...` to just above the code block, think it makes it read better?